[FLINK-16048][avro] Support read/write confluent schema registry avro…

danny0405 opened the pull request:
… data from Kafka

## What is the purpose of the change

Supports read/write with SQL using schema registry avro format.

**The format details**

_The factory identifier (or format id)_

There are 2 candidates now ~

`avro-sr`: the pattern borrowed from KSQL JSON_SR format [1]
`avro-confluent`: the pattern borrowed from Clickhouse AvroConfluent [2]

Personally i would prefer avro-sr because it is more concise and the confluent is a company name which i think is not that suitable for a format name.

**The format attributes**

Options | required | Remark
-- | -- | --
schema-registry.url | true | URL to connect to schema registry service
schema-registry.subject | false | Subject name to write to the Schema Registry service, required for sink

**Note**: the avro schema string is always inferred from the DDL schema, so user should keep the nullability correct (DDL type default is nullable but avro default is non-nullable).

## Brief change log

  - Add `avro-sr` read/write row data format
  - Add tests


## Verifying this change

Added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented

sjwiesman commented on flink-formats/flink-avro-confluent-registry/pom.xml:
```diff
@@ -30,7 +30,7 @@ under the License.
 
 	<properties>
-		<confluent.schema.registry.version>4.1.0</confluent.schema.registry.version>
+		<confluent.schema.registry.version>5.5.1</confluent.schema.registry.version>
```
Update the version in the License notice file

sjwiesman commented:
+1 for SR and please tag me when you open a documentation PR for this feature

danny0405 commented:
> +1 for SR and please tag me when you open a documentation PR for this feature

Sure， thanks for taking care the document.

danny0405 commented:
> After a quick glimpse. Could we unify the `AvroRowDataDeserializationSchema` with `ConfluentRegistryAvroRowDataDeserializationSchema` and `RegistryAvroRowDataSerializationSchema`?
> I really believe we need just a single AvroRowDeserializationSchema for avro for table API.
> 
> I am quite sure sth like this would work:
> 
> ```
> RowDataDeserializationSchema extends ResultTypeQueryable {
>   	private final DeserializationSchema<GenericRecord> nestedSchema;
> 	private final DeserializationRuntimeConverter runtimeConverter;
> 	private final TypeInformation<RowData> resultType;
> 
> 	public AvroRowDataDeserializationSchema2(
> 			DeserializationSchema<GenericRecord> nestedSchema,
> 			DeserializationRuntimeConverter runtimeConverter,
> 			TypeInformation<RowData> resultType) {
> 		this.nestedSchema = nestedSchema;
> 		this.runtimeConverter = runtimeConverter;
> 		this.resultType = resultType;
> 	}
> 
> 	@Override
> 	public void open(InitializationContext context) throws Exception {
> 		nestedSchema.open(context);
> 	}
> 
> 	@Override
> 	public RowData deserialize(byte[] message) throws IOException {
> 		try {
> 			GenericRecord deserialize = nestedSchema.deserialize(message);
> 			return (RowData) runtimeConverter.convert(deserialize);
> 		} catch (Exception e) {
> 			throw new IOException("Failed to deserialize Avro record.", e);
> 		}
> 	}
> 
>        	@Override
> 	public boolean isEndOfStream(RowData nextElement) {
> 		return false;
> 	}
> 
> 	@Override
> 	public TypeInformation<RowData> getProducedType() {
> 		return resultType;
> 	}
> }
> ```
> 
> and then you would use it like this:
> 
> in `AvroFormatFactory`:
> 
> ```
> new RowDataDeserializationSchema(
> 	AvroDeserializationSchema.forGeneric(AvroSchemaConverter.convertToSchema(rowType)),
>         createRowConverter(rowType), // we would need to move this method to some utils class or to a common abstract class for factories
> 	rowDataTypeInfo
> );
> ```
> 
> in `RegistryAvroFormatFactory`:
> 
> ```
> new RowDataDeserializationSchema(
> 	ConfluentRegistryAvroDeserializationSchema.forGeneric(
> 		AvroSchemaConverter.convertToSchema(rowType),
> 		schemaRegistryURL
> 	),
>         createRowConverter(rowType),
> 	rowDataTypeInfo
> );
> ```

Thanks for the nice review, i have addressed your comments.

dawidwys commented on flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/RegistryAvroRowDataSeDeSchemaTest.java:
```diff
@@ -0,0 +1,129 @@
+ * Tests for {@link AvroRowDataDeserializationSchema} and
+ * {@link AvroRowDataSerializationSchema} for schema registry avro.
+ */
+public class RegistryAvroRowDataSeDeSchemaTest {
```
Those tests have nothing to do with schema registry.

They test the same logic as in `AvroRowDataDeSerializationSchemaTest`

danny0405 replied:
Yes, we can remove it.

dawidwys commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroToRowDataConverters.java:
```diff
@@ -0,0 +1,246 @@
+import static org.joda.time.DateTimeConstants.MILLIS_PER_DAY;
+
+/** Tool class used to convert from Avro {@link GenericRecord} to {@link RowData}. **/
+public class AvroToRowDataConverters {
```
`@Internal`

dawidwys commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:
```diff
@@ -169,6 +171,121 @@ private AvroSchemaConverter() {
+	 *
+	 * @return data type matching the schema
+	 */
+	public static DataType convertToDataType(String avroSchemaString) {
```
Could we add tests for this method?

dawidwys commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDataSerializationSchema.java:
```diff
@@ -64,61 +41,59 @@
+	 * Creates an Avro serialization schema with the given record row type.
 	 */
-	private transient Encoder encoder;
+	public AvroRowDataSerializationSchema(RowType rowType) {
```
How about we remove this ctor? IMO the logic from this ctor should be only in the factory. I know that this class in theory is `PublicEvolving` but practically it is only usable from Table API through the factory. Therefore in my opinion it is safe to drop this ctor.

The same applies to `SerializationSchema`.

danny0405 replied:
I would prefer a constructor with default implementation.

dawidwys replied:
Why?

danny0405 replied:
Why not, the avro row data format is default to SE/DE avro without schema, if other formats what to customize something, use the correct constructor. Write the same code everywhere just makes no sense.

dawidwys commented on flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/RegistryAvroRowDataSeDeSchemaTest.java:
```diff
@@ -0,0 +1,129 @@
+ * {@link AvroRowDataSerializationSchema} for schema registry avro.
+ */
+public class RegistryAvroRowDataSeDeSchemaTest {
+	private static final String ADDRESS_SCHEMA = "" +
```
IMO we should add a simple test for serializing and deserializing using schema registry. It does not need to be very in depth, but  so that it checks that everything is well connected.

danny0405 replied:
I have added a schema registry server there, and test the new format SE/DE with connection to the schema registry service. See the new `RegistryAvroRowDataSeDeSchemaTest`.

danny0405 commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/CachedSchemaCoderProvider.java:
```diff
@@ -0,0 +1,76 @@
+/** A {@link SchemaCoder.SchemaCoderProvider} that uses a cached schema registry
+ * client underlying. **/
+@Internal
+class CachedSchemaCoderProvider implements SchemaCoder.SchemaCoderProvider {
```
Personally i prefer normal interface than static inner one, and also we can reuse this data structure.

dawidwys replied:
Where can we reuse it?

danny0405 replied:
We only need one class for the serialize/deserialize schema.

danny0405 commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroFormatFactoryTest.java:
```diff
@@ -0,0 +1,183 @@
+	public ExpectedException thrown = ExpectedException.none();
+
+	@Before
+	public void before() {
```
We can, make them static also works.

KurtYoung commented:
Would `flink-avro-confluent` be a better module name than `flink-avro-confluent-registry`? IMO `registry` has nothing to do with the format itself.

danny0405 commented:
> Would `flink-avro-confluent` be a better module name than `flink-avro-confluent-registry`? IMO `registry` has nothing to do with the format itself.

It depends on how we understand it, the confluent avro format is mainly designed for schema registry.

danny0405 commented:
Schema registry is a terminology and people always calls "schema registry url"[1], the same for "schema registry subject".
[1] https://docs.confluent.io/current/schema-registry/index.html#high-availability-for-single-primary-setup

dawidwys commented on flink-formats/flink-avro-confluent-registry/src/test/java/org/apache/flink/formats/avro/registry/confluent/RegistryAvroRowDataSeDeSchemaTest.java:
```diff
@@ -0,0 +1,199 @@
+	public void testRowDataWriteReadWithCompatibleSchema() throws Exception {
+		testRowDataWriteReadWithSchema(ADDRESS_SCHEMA_COMPATIBLE);
+		// Validates new schema has been registered.
+		assertThat(client.getAllVersions("address-value").size(), is(1));
```
nit: use `SUBJECT`

dawidwys commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/RegistryAvroDeserializationSchema.java:
```diff
@@ -52,13 +53,22 @@
 		this.schemaCoder = schemaCoderProvider.get();
 	}
 
+	public static RegistryAvroDeserializationSchema<GenericRecord> forGeneric(
```
Remove this method? If you prefer to keep it add missing javadoc.

homepy commented on flink-formats/flink-avro-confluent-registry/src/main/resources/META-INF/services/org.apache.flink.table.factories.Factory:
```diff
@@ -0,0 +1,16 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+org.apache.flink.formats.avro.registry.confluent.RegistryAvroFormatFactory
```
This file does not exist in the release jar (https://mvnrepository.com/artifact/org.apache.flink/flink-avro-confluent-registry/1.11.2)， but some other things. 
Maybe there is any mistake of the maven-shade-plugin config in pom.xml?

dawidwys replied:
It is not available, because it will be included in the 1.12 version.

homepy replied:
Thanks. 
I found it...
I need to use the master branch now...

maver1ck commented:
@danny0405 
Any reasons all the fields read and written by this format has prefix 'record_' ?

danny0405 commented:
> @danny0405 @dawidwys
> Any reasons all the fields read and written by this format has prefix 'record_' ? (I'm using flink sql for this client)
> I found responsible code probably here but still have problem with this solution:
> https://github.com/apache/flink/blob/de87a2debde8546e6741390a81f43c032521c3c0/flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java#L365

It's because of the current strategy to infer the Avro schema is convert from the `CREATE TABLE` DDL, and there is no way to get the record name here. So we put a constant `record` as a prefix. The record write out all have explicit field name, but the type should be compatible.

maver1ck commented:
@danny0405 
Problem is that with this strategy I'm unable to read anything from Kafka using Confluent Registry. Example:
I have data in Kafka with following value schema:
```
{
  "type": "record",
  "name": "myrecord",
  "fields": [
    {
      "name": "f1",
      "type": "string"
    }
  ]
}
```
I'm creating table using this avro-confluent format:
```
create table `test` (
	`f1` STRING
) WITH (
  'connector' = 'kafka', 
  'topic' = 'test', 
  'properties.bootstrap.servers' = 'localhost:9092', 
  'properties.group.id' = 'test1234', 
   'scan.startup.mode' = 'earliest-offset', 
  'format' = 'avro-confluent'
  'avro-confluent.schema-registry.url' = 'http://localhost:8081'
);
```
When trying to select data I'm getting error:
```
SELECT * FROM test;
[ERROR] Could not execute SQL statement. Reason:
org.apache.avro.AvroTypeException: Found myrecord, expecting record, missing required field record_f1
```

danny0405 commented:
Thanks, i think this is a bug, i have logged an issue there. See https://issues.apache.org/jira/browse/FLINK-19779

danny0405 commented:
I have fired a fix https://github.com/apache/flink/pull/13763/files, can you help check if possible @maver1ck :)

maver1ck commented:
I will check... mvn is running

maver1ck commented:
OK. It's working. I'm able to read data.

maver1ck commented:
@danny0405 
I think we have one more problem.
When Flink is creating schema in registry nullability is not properly set for logical types.
Examples. Table:
```
create table `test_logical_null` (
	`string_field` STRING,
	`timestamp_field` TIMESTAMP(3)
) WITH (
  'connector' = 'kafka', 
  'topic' = 'test-logical-null', 
  'properties.bootstrap.servers' = 'localhost:9092', 
  'properties.group.id' = 'test12345', 
   'scan.startup.mode' = 'earliest-offset', 
  'format' = 'avro-confluent', -- Must be set to 'avro-confluent' to configure this format.
  'avro-confluent.schema-registry.url' = 'http://localhost:8081', -- URL to connect to Confluent Schema Registry
  'avro-confluent.schema-registry.subject' = 'test-logical-null' -- Subject name to write to the Schema Registry service; required for sinks
)
```
Schema:
```
{
  "type": "record",
  "name": "record",
  "fields": [
    {
      "name": "string_field",
      "type": [
        "string",
        "null"
      ]
    },
    {
      "name": "timestamp_field",
      "type": {
        "type": "long",
        "logicalType": "timestamp-millis"
      }
    }
  ]
}
```
For not null fields:
```
create table `test_logical_notnull` (
	`string_field` STRING NOT NULL,
	`timestamp_field` TIMESTAMP(3) NOT NULL
) WITH (
  'connector' = 'kafka', 
  'topic' = 'test-logical-notnull', 
  'properties.bootstrap.servers' = 'localhost:9092', 
  'properties.group.id' = 'test12345', 
   'scan.startup.mode' = 'earliest-offset', 
  'format' = 'avro-confluent', -- Must be set to 'avro-confluent' to configure this format.
  'avro-confluent.schema-registry.url' = 'http://localhost:8081', -- URL to connect to Confluent Schema Registry
  'avro-confluent.schema-registry.subject' = 'test-logical-notnull-value' -- Subject name to write to the Schema Registry service; required for sinks
);
```
Schema
```
{
  "type": "record",
  "name": "record",
  "fields": [
    {
      "name": "string_field",
      "type": "string"
    },
    {
      "name": "timestamp_field",
      "type": {
        "type": "long",
        "logicalType": "timestamp-millis"
      }
    }
  ]
}
```
As you can see for string_field we have proper union with null (for nullable field). For timestamp_field in both examples union is missing.

danny0405 commented:
@maver1ck , you are right, we ignore the nullability of `TIMESTAMP_WITHOUT_TIME_ZONE`, `DATE`, and `TIME_WITHOUT_TIME_ZONE` and `Decimal`, would fix it altogether in this PR.

maver1ck commented:
Thanks @danny0405

danny0405 commented:
I have updated the fix, @maver1ck , please check if you have time.

maver1ck commented:
@danny0405 I see code review is still in progress.
Could you please let me know when this PR would be polished a little ?
